Proxy: Dedicated DaemonSet for Envoy Proxy

mhofstetter opened the pull request:
Currently, whenever L7 capabilities are required, the Cilium L7 Proxy (Envoy) gets executed as separate process within the CIlium Agent k8s container.

This PR adds support for deploying Envoy independently as a separate K8s DaemonSet for availability, performance, and (potentially ) security benefits.

**Potential Benefits**

* Restarting cilium-agent for upgrades does not impact live dataplane traffic (this is a property of cilium in general for the BPF datapath, but no longer applies once any Envoy-based forwarding is brought in).
* Failures of cilium-agent to start that result in cilium pods in crashloop backoff do not impact dataplane traffic.
* Performance isolation between Envoy and cilium-agent. An overloaded Envoy will not swamp the cilium-agent, or vice versa. Can also have separate CPU/mem limits.

**Tasks**

- [x] Separate directory `/var/run/cilium/envoy/sockets` for Envoy proxy sockets (admin, xDS, accesslogger) -> will be shared via host between agent & Envoy
- [x] Agent Configuration (`external-envoy-proxy`), whether envoy is deployed in DaemonSet or whether it should be started in embedded mode (defaults to false for the time being)
- [x] Introduce CIlium Envoy DaemonSet in Cilium Helm Chart
- [x] Using Image `quay.io/cilium/cilium-envoy` from https://github.com/cilium/proxy (which became an ubuntu runtime for a start)
- [x] Bugtool: adapt changed admin api socket path when gathering Envoy Config during Cilium SysDump
- [x] Add Metrics Listener as static resource when deploying Envoy  as DaemonSet
- [x] Use `/ready` [endpoint](https://www.envoyproxy.io/docs/envoy/latest/operations/admin#get--ready) as health endpoint for Envoy Pods
- [x] Envoy Version check via Admin Interface instead of checking via binary version output
- [x] Configurable Envoy application log configuration (log path, log format, log level) -> Envoy Application Log will only be visible in CIlium Envoy container
- [x] Temporary GitHub action based on `Conformance Kind` for testing Cilium connectivity tests against cilium with Envoy in DaemonSet (based on latest Cilium CLI to be able to use new helm mode) 
- [x] Add Envoy deployment mode to Cilium proxy status (CLI)

This PR tries to cover the relevant tasks to provide support for deploying Envoy in its own DaemonSet - without getting too big. The following tasks were out of scope.

* Optimize runtime base image of Envoy (currently still builds on top of the same base image as the agent)
* Reduction of the permission of the Agent DaemonSet deployment
* Resource configuration for the Envoy DaemonSet deployment
* Optimized behaviour if Envoy isn't deployed
* proxylib integration (https://github.com/cilium/cilium/issues/25226)
* OpenShift support (communication without unix sockets)
* ...

Please review the individual commits! They are more or less aligned with above tasks.

**Installation**

For the time being, the default will still be the "embedded" Envoy proxy. Therefore, Envoy as DaemonSet must be explicitly enabled via helm value `envoy.enabled=true`.

* Helm: supported
* Cilium CLI: out of the box supported via newly introduced "helm mode" (requires latest Cilium CLI from main branch).

:information_source: Adaptions (status, sysdump, install) in [cilium/cilium-cli](https://github.com/cilium/cilium-cli) will follow soon (edit: https://github.com/cilium/cilium-cli/pull/1546)

**Release Note**

```release-note
Support for deploying Cilium L7 Proxy (Envoy) independently as a separate DaemonSet for availability, performance, and security benefits.
```

mhofstetter committed:
envoy proxy: separate directory for envoy proxy sockets

Currently, the /var/run/cilium is used as state and runtime directory
for various aspects. The Unix Sockets for the communciation between
Cilium Agent & Ciliums Envoy (xDS, Admin, AccessLog) are created
in this directory too.

With the need of sharing the sockets between Agent & Envoy Pod via
hostmount in the future, this commit moves the sockets into the
directory /var/run/cilium/envoy/sockets.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
helm: add Cilium Envoy DaemonSet to Cilium Helm Chart

This commit introduces the DaemonSet for the Cilium Envoy.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
envoy proxy: introduce configuration for starting Envoy in embedded mode

Currently, the Cilium Envoy Proxy is started as standalone process embedded
within the Cilium Agent Pod.

During the transition towards having a dedicated DaemonSet for the
Cilium Envoy proxy, this commit introduces the possibility to configure
whether the embedded proxy should be enabled or not via Cilium config
property `external-envoy-proxy`.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
envoy proxy: envoy admin client refactoring

This commit refactors the initialization of the embedded Envoy Proxy
instance by introducing a EnvoyAdminClient struct which encapsulates
the communication with the Envoy Proxy instance - whether this is the
embedded or standalone (DaemonSet) instance.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
bugtool: change socket path for envoy admin socket

This commit adapts the change of the envoy unix socket location when
dumping the Envoy config & metrics. Unix Sockets are now located within
`/var/run/cilium/envoy/sockets`.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
helm & envoy proxy: metrics listener as static resource for Envoy DaemonSet

Currently, the metrics listener is added dynamically at runtime - mainly
to prevent the embedded Envoy from not starting if the port is already
used.

In case of an dedicated Envoy DaemonSet, the metrics listener is gets
configured as static resource witthin the bootstrap config.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
helm: use /ready from Envoy admin iface for healthprobes on daemonset

With this commit, the Envoy DaemonSet uses the "/ready" from the Envoy
admin interface for health probes (startup, live & ready).

Therefore, an additional listener (default: 9878) gets added to Envoy.
Reusage of prometheus listener is prevented to keep the things separated
and provide the flexiblity which might be needed for prometheus.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
envoy proxy: Envoy version check via admin interface during initialization

Currently, the Envoy version of the Cilium proxy gets checked at startup
via the binaries version command.

Due to extraction of the Cilium Envoy Proxy into its own DaemonSet, this
commits retrieves the version via Envoys admin interface. Therefore,
version check happens after initializing the EnvoyAdminClient.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
helm: Adapt envoy log path for Envoy DaemonSet

Currently, the embedded Envoy Proxy instance supports changing the log
path from stdout to a file via cilium config property `envoy-log`.

This commit introduces the Helm property `envoy.log.path` which
configures the Cilium property and the Envoy DaemonSet start arguments.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
helm: Configurable Envoy log format

Currently, the embedded Envoy Proxy instance hardcodes Envoys
application log format to an adapted form of the Cilium log format and
switches to the default Envoy log format if a log path is defined.

By having the Cilium Envoy Proxy in a dedicated DaemonSet, the default is the
Envoy log format.

In addition, this commit provides the possibiliity to configure the log
format via Helm value `envoy.log.format`.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
helm: Adapt envoy log level to Envoy DaemonSet

This commit adapts Envoys Log Level in the Envoy DaemonSet to use the
debug & debug-verbose options from the Helm Chart.

* "debug.verbose" contains "envoy" -> trace
* "debug.enabled" & "debug.verbose" contains "flow" -> debug
* default -> info

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
envoy proxy: move embedded envoy aspects into dedicated file

This commit refactors all aspects of the embedded Envoy proxy instance
into a separate file `embedded_envoy.go` and separates it from other
aspects like EnvoyAdminClient.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
ci: temporary GHA Conformance Kind test with envoy daemonset

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

jrajahalme commented on images/proxy/Dockerfile:
```diff
@@ -0,0 +1,30 @@
+# Copyright Authors of Cilium
+# SPDX-License-Identifier: Apache-2.0
+
+ARG CILIUM_RUNTIME_IMAGE=quay.io/cilium/cilium-runtime:b32efb0e682e74311b628df328dfd768108b92c2@sha256:0873607c0fc991c884bc7d39ea774c8ebd7df2be35f4b69b6d16b33ce871427b
```
Once we figure out the minimal runtime base for this image, maybe we should make `cilium/proxy` build that image by default, instead of the current image containing only the `cilium-envoy` binary. One less image, and the `cilium-envoy` binary could be copied from that for the "built-in" version like today?

mhofstetter replied:
good point. maybe in combination with "reducing" the base runtime image of envoy. then we could just build on top of this image when building the quay.io/cilium/proxy image in the cilium/cilium repo (which still makes sense to keep the versioning aligned with cilium/cilium) - instead of just copying files.

jrajahalme commented on pkg/defaults/defaults.go:
```diff
@@ -41,6 +41,9 @@ const (
 	RuntimePath = "/var/run/cilium"
 
+	// ProxySocketPath is the default path to the proxy socket directory
+	ProxySocketPath = "/var/run/proxy"
```
IMO it would be good if this would still be under `/var/run/cilium`, maybe `/var/run/cilium/proxy`?

jrajahalme replied:
This would also simplify many of the changes in the PR, as we could build proxy socket path from runtime path + "/proxy". We already do similar pattern for other subdirs in the runtime dir, I believe.

mhofstetter replied:
will have a look. should work. at the end we just need a separate directory for the sockets and not mixing it with other stuff. otherwise its not really possible to mount it to the host.

jrajahalme replied:
I don't yet understand the aspect that it needs to be a separate directory for just the sockets? Why could the bootstrap, for example be in the same directory?

mhofstetter replied:
@jrajahalme 

because the sockets directory needs to be mounted as a whole to the host via hostpath volume - so envoy & agent can communicate with each other.

this would clash with the bootstrapconfig which gets mounted via configmap volumemount.

mhofstetter replied:
@jrajahalme 

changed the default proxy socket path to `/var/run/cilium/envoy/sockets` - by building on top of the runDir/stateDir.

the bootstrapconfig gets generated (embedded) or mounted (daemonset - from configmap) into `/var/run/cilium/envoy`.

jrajahalme commented on pkg/option/config.go:
```diff
@@ -431,6 +431,9 @@ const (
 	ProxySocketDir = "proxy-socket-dir"
 
+	// EmbeddedProxyEnabled defines whether the embedded Cilium Proxy should be enabled or not.
+	EmbeddedProxyEnabled = "proxy-embedded-enabled"
```
We also have DNS proxy, so the fast read of this option could be understood to cover both the embedded DNS proxy and the embedded Envoy proxy. If the intent is to not cover the DNS proxy here, then we should name this e.g. `embedded-envoy-enabled` instead.

jrajahalme replied:
This also relates to a larger uncertainty I have on the naming. IMO it would be nice to have the container named as `cilium-envoy` to give an outward acknowledgement that we are using Envoy, and that we are not trying to hide it behind `cilium-proxy`. I know this deviates from the github repo naming, but so does `cilium-agent` from `cilium/cilium`.

mhofstetter replied:
tend to use the proposal of @sayboras and go with `external-envoy-proxy` - with the additional benefit of naming it after the mode which eventually will survive (even after embedded mode has been deleted)

mhofstetter replied:
@jrajahalme @sayboras changed to use `external-envoy-proxy`

sayboras replied:
:100:

jrajahalme commented on install/kubernetes/cilium/templates/cilium-proxy/daemonset.yaml:
```diff
@@ -0,0 +1,258 @@
+      hostNetwork: true
+      {{- if and .Values.etcd.managed (not .Values.etcd.k8sService) }}
+      # In managed etcd mode, Cilium must be able to resolve the DNS name of
+      # the etcd service
```
`cilium-proxy` is not accessing etcd, so this part is necessary.

mhofstetter replied:
thanks. deleted this part with keeping the possibility of a configurable dnsPolicy itself (`dnsPolicy: {{ .Values.proxy.dnsPolicy }}`)

jrajahalme commented on install/kubernetes/cilium/templates/cilium-proxy/daemonset.yaml:
```diff
@@ -0,0 +1,258 @@
+              path: envoy-bootstrap-config.json
+        # To keep state between restarts / upgrades
+      {{- if and .Values.bpf.autoMount.enabled }}
+        # To keep state between restarts / upgrades for bpf maps
```
`cilium-envoy` will still need to be restarted if Cilium upgrade/downgrade re-creates the conntrack and/or ipcache maps due to changes in the map key/value size or layout. Currently this information is built in to `cilium-envoy`, but maybe we could:
- make sure the maps are re-opened whenever Cilium has restarted
  - it should be possible to detect this from the xds stream being restarted
- probe the bpf map key value sizes, and assume the fields we need remain in fixed offsets, or also probe this (vie BTF?)

mhofstetter replied:
@jrajahalme thanks, good point. i assume this could be a candidate which could be implemented in a separate PR?

jrajahalme replied:
Yes, definitely good for a separate PR, this is a thorny issue all by itself.

jrajahalme commented on install/kubernetes/cilium/templates/cilium-proxy/service.yaml:
```diff
@@ -0,0 +1,48 @@
+    protocol: TCP
+    targetPort: envoy-metrics
+{{- else if .Values.proxy.prometheus.enabled }}
+apiVersion: v1
```
Most of this is a duplicate from above, would it be possible to eliminate this duplication?

sayboras replied:
agree with the above, can we just guard annotation block based on `.Values.proxy.prometheus.serviceMonitor.enabled` ?

mhofstetter replied:
thanks. refactored this part so the annotations are only added if servicemonitor isn't enabled and detection is done by analyzing annotations.

jrajahalme commented on install/kubernetes/cilium/templates/cilium-proxy/servicemonitor.yaml:
```diff
@@ -0,0 +1,39 @@
+apiVersion: monitoring.coreos.com/v1
+kind: ServiceMonitor
+metadata:
+  name: cilium-agent
```
This seems like possible duplication of cilium-agent's ServiceMonitor?

mhofstetter replied:
yep, renamed to `cilium-proxy`

jrajahalme commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -1688,10 +1693,187 @@ dashboards:
+    # Socket directory within the proxy & agent containers
+    container: "/var/run/proxy"
+    # Socket directory within the host
+    host: "/var/run/cilium-proxy"
```
Could this be `/var/run/cilium/proxy`?

mhofstetter replied:
@jrajahalme 

for the sake of simplicity in both cases `/var/run/cilium/envoy/sockets` is used. IMO it only makes sense to change the one of the host anyway.

jrajahalme commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -1688,10 +1693,187 @@ dashboards:
+  # Location of unix sockets for communication between Agent & Proxy.
+  socketDir:
+    # Socket directory within the proxy & agent containers
+    container: "/var/run/proxy"
```
Could this be `/var/run/cilium`? It is for contacting Cilium Agent, right?

mhofstetter replied:
@jrajahalme 

for the sake of simplicity in both cases (actually 3: agent container, envoy container, host) `/var/run/cilium/envoy/sockets` is used. IMO it only makes sense to change the one of the host anyway.

and yes it's for the communication between agent <-> envoy: 

agent -> envoy: admin
envoy -> agent: xDS & accesslogger

jrajahalme commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -1688,10 +1693,187 @@ dashboards:
+  image:
+    override: ~
+    repository: "${CILIUM_PROXY_REPO}"
+    tag: "${CILIUM_VERSION}"
```
This will need to be separated from `CILIUM_VERSION`, right?

mhofstetter replied:
why? i would keep it aligned with the version of cilium itself - like all the other components do (operator, hubble relay, clustermesh apiserver, ...)

jrajahalme replied:
Eventually we'll want to be able to update Cilium agent independently of the Envoy, implying the break-up of the tight versioning between them. cilium/proxy versioning does not currently follow cilium versioning, but Envoy versioning.

mhofstetter replied:
@jrajahalme yep, eventually. but as discussed probably not with this PR :) therefore, currently this `CILIUM_VERSION` should be correct here?

mhofstetter replied:
@jrajahalme 

PR is now using the image `quay.io/cilium/cilium-envoy` (with added ubuntu runtime) as its image.

mhofstetter replied:
@jrajahalme @sayboras rebased to `main` to resolve conflicts with envoy version.

now using the cilium-envoy image with the envoy patch version included in `images/cilium/Dockerfile` & `install/kubernetes/cilium/values.yaml.tmpl`

jrajahalme commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -1691,10 +1696,187 @@ dashboards:
+  image:
+    override: ~
+    repository: "quay.io/cilium/proxy"
+    tag: "latest"
```
tag will need to be a real proxy image tag?

mhofstetter replied:
this seems to be correct. at least all the other images have tag `latest` too. this file is generated from the template anyway. will be overwritten with a stable release version in most of the cases anyway (by surrounding tooling)

mhofstetter replied:
@jrajahalme 

PR is now using the image `quay.io/cilium/cilium-envoy` (with added ubuntu runtime) as its image.

jrajahalme reviewed (commented):
Awesome work, thanks for taking this on!

I'd like to see some discussion on some of the naming aspects, on cilium-agent command line option, as well as on the new container name and the paths used for communication between the Envoy proxy container and Cilium Agent.

sayboras commented on images/proxy/Dockerfile:
```diff
@@ -0,0 +1,30 @@
+ARG TARGETARCH
+LABEL maintainer="maintainer@cilium.io"
+RUN echo ". /etc/profile.d/bash_completion.sh" >> /etc/bash.bashrc
+COPY --from=cilium-envoy / /
```
do we need to copy the proxylib over as well?

mhofstetter replied:
@sayboras we're not doing this for the embedded mode in the agent image either. or am i missing something?

sayboras replied:
the file is libcilium.so, which is built from https://github.com/cilium/cilium/blob/main/proxylib/Makefile. In cilium/proxy, we have a step to manually copy this file to run the test.

https://github.com/cilium/proxy/blob/main/.github/workflows/ci-tests.yaml#L40

mhofstetter replied:
@sayboras 

thanks for the pointer. 

if we need to add this to the image in the cilium/cilium build we can't need the image `quay.io/cilium/cilium-envoy` from cilium/proxy as is :thinking: 

will have a look and contact you next week

mhofstetter replied:
will be solved in a subsequent PR

sayboras commented on daemon/cmd/daemon_main.go:
```diff
@@ -735,6 +735,9 @@ func initializeFlags() {
 	flags.String(option.ProxySocketDir, defaults.ProxySocketPath, "Directory path where proxy sockets are located")
 	option.BindEnv(Vp, option.ProxySocketDir)
 
+	flags.Bool(option.EmbeddedProxyEnabled, true, "Whether the embedded Cilium Proxy should be enabled or not")
```
what do you think about external-envoy-proxy ?

mhofstetter replied:
agree, will change that. naming it this way also have the advantage that we can keep it even after deleting the embedded mode.

mhofstetter replied:
@sayboras changed to `external-envoy-proxy`

sayboras commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -1688,10 +1693,187 @@ dashboards:
+    rollingUpdate:
+      maxUnavailable: 2
+  # -- Roll out cilium proxy pods automatically when configmap is updated.
+  rollOutOnConfigMapUpdate: false
```
nit: rollOutOnConfigMapUpdate -> rollOutPods for consistency with other components.

mhofstetter replied:
ok, thought of naming this "correctly". but changed it to be consistent with the others :)

mhofstetter replied:
@sayboras renamed

sayboras commented on install/kubernetes/cilium/templates/cilium-proxy/configmap.yaml:
```diff
@@ -0,0 +1,12 @@
+  name: cilium-proxy-config
+  namespace: {{ .Release.Namespace }}
+data:
+{{- (tpl (.Files.Glob "files/cilium-proxy/configmap/*").AsConfig .) | nindent 2 }}
```
multiple files might not work here, so we can just explicitly add filename here.

mhofstetter replied:
it would work, but might not be what's expected. set the filename of the bootstrap config explicitly.

sayboras commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -1688,10 +1693,187 @@ dashboards:
+    host: "/var/run/cilium-proxy"
+
+  # Time in seconds after which a TCP connection attempt times out
+  connectTimeout: 2
```
as the attribute name didn't contain unit (e.g. connectTimeoutSeconds), we normally have value as 2s instead. This will help to remove `s` append in bootstrap json file.

mhofstetter replied:
these config properties already exist in the cilium agent. i just added them to the helm chart to be able to pass the same values to the bootstrapconfig and the agent configmap.

```
  proxy-connect-timeout: {{ .Values.proxy.connectTimeout | quote }}
  proxy-max-requests-per-connection: {{ .Values.proxy.maxRequestsPerConnection | quote }}
  proxy-max-connection-duration-seconds: {{ .Values.proxy.maxConnectionDuration | quote }}
```

and `proxy-connect-timeout` already doesn't contain the "in seconds" in its name. therefore if i change something, i tend to fix the name in the helm chart and add `Seconds` at the end. would this be ok for you? this way we still have to add the `s` as suffix in the bootstrap - but we have to add or remove it somewhere (if we don't want to rename the existing cilium config property or change its type)

mhofstetter replied:
@sayboras added suffix `Seconds` to `connectTimeout` & `maxConnectionDuration`

sayboras commented on install/kubernetes/cilium/files/cilium-proxy/configmap/envoy-bootstrap-config.json:
```diff
@@ -0,0 +1,197 @@
+      {
+        "name": "egress-cluster-tls",
+        "type": "ORIGINAL_DST",
+        "connectTimeout": "{{ .Values.proxy.connectTimeout }}s",
```
we can remove suffix `s` here, and let the value have unit (e.g. 20s) instead.

sayboras commented on install/kubernetes/cilium/files/cilium-proxy/configmap/envoy-bootstrap-config.json:
```diff
@@ -0,0 +1,197 @@
+          "envoy.extensions.upstreams.http.v3.HttpProtocolOptions": {
+            "@type": "type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions",
+            "commonHttpProtocolOptions": {
+              "maxConnectionDuration": "{{ .Values.proxy.maxConnectionDuration }}s",
```
same as above, the duration in naming implies that value can have unit (e.g. s, m, etc)

sayboras commented on install/kubernetes/cilium/files/cilium-proxy/configmap/envoy-bootstrap-config.json:
```diff
@@ -0,0 +1,197 @@
+            "useDownstreamProtocolConfig": {}
+          }
+        },
+        "cleanupInterval": "{{ .Values.proxy.connectTimeout }}.500s"
```
same as above.

sayboras reviewed (commented):
Great stuffs, I have a couple of comments as per below.

mhofstetter committed:
envoy proxy: generate bootstrap.pb into /var/run/cilium/envoy

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
status/cli: add envoy deployment mode to proxy status

With this commit, the Envoy deployment mode (embedded or external
(as DaemonSet)) gets added to the Cilium Proxy status.

```
...
Proxy Status:            OK, ip 10.0.1.172, 6 redirects active on ports 10000-20000, Envoy: external
...
```

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

sayboras commented on install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:
```diff
@@ -234,10 +234,10 @@ spec:
           containerPort: {{ .Values.prometheus.port }}
           hostPort: {{ .Values.prometheus.port }}
           protocol: TCP
-        {{- if .Values.proxy.prometheus.enabled }}
```
ideally we would like to keep old attribute as deprecated for upgrade path.

mhofstetter replied:
@sayboras 

yep - indeed. refactored all occurrences of `.Values.envoy.prometheus.enabled` & `.Values.envoy.prometheus.port` to use the old "proxy value" instead of the new one.

in addition i documented the old values as deprecated.

this way we should be safe if someone is configuring and therefore passes the old values.

aanm replied:
@mhofstetter this change needs to be documented in the upgrade guide.

sayboras commented on install/kubernetes/cilium/templates/cilium-envoy/daemonset.yaml:
```diff
@@ -0,0 +1,254 @@
+        args:
+        - '-c /var/run/cilium/envoy/bootstrap-config.json'
+        - '--base-id 0'
+        {{- if and (hasKey .Values.debug "verbose") ( has "envoy" .Values.debug.verbose) }}
```
I got the below error with flags --set debug.enabled=true --set debug.verbose=envoy.

```
stderr: "Error: template: cilium/templates/cilium-envoy/daemonset.yaml:68:54: executing \"cilium/templates/cilium-envoy/daemonset.yaml\" at <has \"envoy\" .Values.debug.verbose>: error calling has: Cannot find has on type string\n\nUse --debug flag to render out invalid YAML\n"
```

mhofstetter replied:
@sayboras 

thanks. i assumed that this is an array and not a comma-separated list :man_facepalming: 

anyway - fixed it. it can handle commaseparated values in the `debug.verbose` parameter.

sayboras commented on pkg/envoy/envoy.go:
```diff
@@ -134,139 +135,141 @@ func GetEnvoyVersion() string {
+	// not meaningful.
+	// Not using the normal localhost address to make it obvious that we are not
+	// connecting to Envoy's admin interface via the IP stack.
+	adminAddress := "192.0.2.34:56"
```
can you use some host name/alias like envoy-admin here ?

mhofstetter replied:
@sayboras 

yes it is possible. got rid of it (even though the author wrote it this way on purpose) - but let the necessary parts of the comment there.

sayboras commented on install/kubernetes/cilium/templates/cilium-envoy/service.yaml:
```diff
@@ -0,0 +1,27 @@
+    app.kubernetes.io/part-of: cilium
+    io.cilium/app: proxy
+spec:
+  clusterIP: None
```
any reason to use headless service here ?

mhofstetter replied:
@sayboras 

assumption: without using the CRD `ServiceMonitor`, this service is used by the prometheus operator to find all pods which prometheus should scrape.

Therefore it needs the headless service to resolve to all pods when resolving the service via DNS - instead of actually making loadbalancing.

squeed replied:
Indeed, this is more a "bag of pods" than a service.

sayboras commented on install/kubernetes/cilium/files/cilium-envoy/configmap/bootstrap-config.json:
```diff
@@ -5,7 +5,6 @@
     "listeners": [
-      {{- if and .Values.prometheus.enabled .Values.envoy.prometheus.enabled }}
       {
         "name": "envoy-prometheus-metrics-listener",
```
we can rename this listener to something more generic, as now /healthz endpoint is part of this listener.

mhofstetter replied:
@sayboras 

yep this is true. i let the name unchanged because it would change the statsprefix as well - and would break some stuff.

but this point and your [comment below](https://github.com/cilium/cilium/pull/25081#discussion_r1180365425) led me to the conclusion that it might be better to have a separate envoy listener for health at all.

this way we keep the flexibility and have clear configuration (e.g. port) for the two aspects - with the drawback of having an additional port.

-> implemented it this way

sayboras commented on install/kubernetes/cilium/templates/cilium-envoy/daemonset.yaml:
```diff
@@ -75,11 +75,8 @@ spec:
             host: {{ .Values.ipv4.enabled | ternary "127.0.0.1" "::1" | quote }}
             path: /healthz
-            port: {{ .Values.envoy.healthPort }}
+            port: {{ .Values.envoy.prometheus.port }}
```
this implied prometheus.enabled must be true, however, it seems like we always need health check port.

mhofstetter replied:
@sayboras 

thanks - yes that's irritating. 

re-used the `healthPort` when implemented the dedicated envoy listener for health aspects - as mentioned in [above comment](https://github.com/cilium/cilium/pull/25081#discussion_r1180634222).

sayboras reviewed (commented):
Couple of comment as below, the rest looks good to me :+1:

mhofstetter committed:
envoy proxy: replace pseudo ip with envoy-admin in EnvoyAdminClient

Currently, when accessing the Envoy Admin API via Unix domain socket,
the host is set to an arbitrary IP & port. The intention is to prevent
the assumption that this is a call vai IP stack.

This commit removes the IP with the host `envoy-admin` and leaves the
comment for further information.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
helm: deprecate proxy.prometheus in favor of envoy.prometheus

This commit marks `proxy.prometheus.enabled` & `proxy.prometheus.port`
as deprecated in favor of `envoy.prometheus.enabled` &
`envoy.prometheus.port`.

The implementation will still respect the deprecated values prioritized
if explicitly passed to helm.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
helm: decouple operator & envoy prometheus config from agent

Currently, some values from operator & envoy regarding prometheus depend
on whether agent prometheus is enabled or not.

This commit changes this, so all components are independent from each
other.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

sayboras commented on .github/workflows/conformance-kind-proxy-daemonset.yaml:
```diff
@@ -0,0 +1,141 @@
+name: ConformanceKindEnvoyDaemonSet
```
What do you think about using GHA matrix, and just re-use the same ConformanceKind GHA? Basically, it's to reduce the maintenance effort.

mhofstetter replied:
@sayboras 

thought about that as well. but when introducing this - cilium cli 0.14 (with required helm mode) wasn't released - and therefore the diff even bigger.

and even now it kind of depends on https://github.com/cilium/cilium/issues/25156. 

happy to combine them in an additional step - where we might want to consider other tests too (e.g. ingress & gateway conformance tests).

sayboras reviewed (approved):
> addressed your comments except for the proxylib part (need some more time & probably discussions for that next week)

I am happy to tackle this in subsequent PR, other changes are looking good to me. Great work :medal_military:

mhofstetter committed:
docs: document envoy as daemonset changes

This commit documents the following things:

* new cilium envoy health listener port
* changed cilium envoy prometheus port `proxy.prometheus` ->
  `envoy.prometheus`
* added deprecation of `proxy.prometheus.port` &
  `proxy.prometheus.enabled` to upgrade guide

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

sayboras reviewed (approved):
LGTM, a few follow-up actions such as proxylib, and upgrade docs can be done in subsequent PRs.

sayboras commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -1699,13 +1704,204 @@ dashboards:
+# -- Configure Cilium Envoy options.
+envoy:
+  # Enable Envoy Proxy in standalone DaemonSet.
+  enabled: false
```
I forgot mentioning about one more point, the current helm chart is having .Values.l7proxy flag, which might overlap with what we have here. WDYT?

https://github.com/cilium/cilium/blob/db3e0152c6583e0cbec1013e514526adf7229faf/install/kubernetes/cilium/values.yaml#L1496

mhofstetter replied:
@sayboras 

thanks for pointing that out. but i think it's ok to have `l7Proxy` & `envoy`

* `l7Proxy` is not only about Envoy (http, ...) but also about FQDN
* `l7Proxy` configures the l7 policy aspect within the agent, whereas `envoy` is configuring the deployment of Envoy

bimmlerd commented on pkg/envoy/accesslog_server.go:
```diff
@@ -127,6 +123,13 @@ func (s *accessLogServer) accessLogger(conn *net.UnixConn) {
 }
 
+// isEOF returns true if the error message ends in "EOF". ReadMsgUnix returns extra info in the beginning.
+func isEOF(err error) bool {
```
Does `errors.Is(err, io.EOF)` not work?

mhofstetter replied:
@bimmlerd 

thanks. i just moved this function - but it's possible to replace it - the errors include unwrapping.

bimmlerd commented on api/v1/models/proxy_status.go:
```diff
@@ -54,6 +64,48 @@ func (m *ProxyStatus) Validate(formats strfmt.Registry) error {
+}
+
+const (
+
```
nit: spurious newline

mhofstetter replied:
@bimmlerd 

`Code generated by go-swagger; DO NOT EDIT.`

bimmlerd commented on contrib/testing/kind-values.yaml:
```diff
@@ -11,6 +11,8 @@ operator:
   nodeSelector:
     kubernetes.io/os: linux
     kubernetes.io/hostname: kind-worker
+envoy:
```
nit: isn't this just the default?

mhofstetter replied:
@bimmlerd 

yes, previously there were local image tags of the envoy image as well.

-> deleted the commit

bimmlerd commented on pkg/envoy/envoy.go:
```diff
@@ -68,13 +68,13 @@ func mapLogLevel(level logrus.Level) string {
 }
 
-type admin struct {
+type EnvoyAdminClient struct {
```
this doesn't quite match the commit description - EnvoyAdminClient is still a struct here

mhofstetter replied:
@bimmlerd 

thanks - renamed in the commit message

bimmlerd commented on pkg/proxy/envoyproxy.go:
```diff
@@ -45,9 +47,46 @@ func initEnvoy(runDir string, xdsServer *envoy.XDSServer, wg *completion.WaitGro
 
+func checkEnvoyVersion() error {
+	const versionRetryAttempts = 20
+	const versionRetryWait = 500 * time.Millisecond
```
is there no way of making this event based instead of waiting for a specific timeout? I'm worried that this will (if even enabled) cause flakes in CI due to dependence on timing

mhofstetter replied:
@bimmlerd 

the version check feature is enabled by default - so we would see flakes if this becomes a problem.

changing it to something "event based" would only help in case the flakyness would occur within the retry wait duration - which is quite low with 500ms.

beside this, the whole version check is subject of discussion because we kind of want to support a broader range of envoy versions in the future.

bimmlerd replied:
> so we would see flakes if this becomes a problem

That's what I'm worried about, not super keen on introducing more flakes.

> changing it to something "event based" would only help in case the flakyness would occur within the retry wait duration - which is quite low with 500ms.

I don't follow. I read it as if spawning envoy takes more than 10s (20 * 500ms) then we error out - that seems like something which is in the realm of plausibility for CI, right?

> beside this, the whole version check is subject of discussion because we kind of want to support a broader range of envoy versions in the future.

Sure, maybe it's irrelevant, I wouldn't know :)

mhofstetter replied:
actually, we don't error out - we're only writing an error message to the log. and in most of the cases envoy starts faster than the envoy (within ~1s) - only the embedded case where the check gets executed immediately after the start of envoy might need 1, 2 retries. -> we could easily move it into its own go routing if necessary.

bimmlerd replied:
:facepalm: I apparently couldn't read two lines up

alright, fine by me

bimmlerd reviewed (commented):
Tried to review, but this PR is just too massive. Would it be possible to split it up into multiple PRs? See e.g. Joe's series of PRs culminating in https://github.com/cilium/cilium/pull/24968, which makes the review much easier.

Left some comments inline as far as I got

aanm commented on pkg/envoy/envoy.go:
```diff
@@ -152,115 +152,121 @@ func StartEnvoy(runDir, logPath string, baseID uint64) *Envoy {
+	} else {
+		// Case Envoy as DaemonSet
 		return e
 	}
```
Small nit but if we did if externalEnvoyProxy { return e } in the beginning we could have avoided such a big diff.

mhofstetter replied:
@aanm 

good point - thanks. i refactored it on this commit to ease reviewing for upcoming reviewers -  even though that part got refactored and moved anyway in the context of the whole PR.

aanm commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -1771,7 +1771,7 @@ envoy:
 
   # -- TCP port for the health API.
-  healthPort: 9879
+  healthPort: 9878
```
This needs to be documented in the list of ports that we open on the host.

aanm replied:
This is done in the second to last or last commit.

mhofstetter replied:
@aanm 

yep that one should be covered

aanm commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -1704,10 +1704,12 @@ dashboards:
   prometheus:
+    # -- Deprecated in favor of envoy.prometheus.enabled
     enabled: true
-    port: "9964"
```
Needs upgrade note as well.

mhofstetter replied:
@aanm 

thanks. added the deprecation of `proxy.prometheus.enabled` & `proxy.prometheus.port` in favor of their counterparts in `envoy.prometheus.*` to the upgrade guide.

aanm reviewed (approved):
LGTM, only some small notes that don't require a 2nd review unless requested.

bimmlerd commented on daemon/cmd/daemon_main.go:
```diff
@@ -733,6 +733,9 @@ func initializeFlags() {
 	flags.String(option.StateDir, defaults.RuntimePath, "Directory path to store runtime state")
 	option.BindEnv(Vp, option.StateDir)
 
+	flags.Bool(option.ExternalEnvoyProxy, false, "whether the Envoy is deployed externally in form of a DaemonSet or not")
```
I'm sure you've considered it, but adding stuff to option.Config is mildly against what the modularization effort is working for. There isn't a way to make this a cell?

mhofstetter replied:
@bimmlerd i definitely considered that. but it's not really possible without modularizing the proxy itself.

i try to come back to that :)

bimmlerd reviewed (commented):
I'm not going to block the PR on it, but I disagree that having everything in one PR is better due to context. The required context can be carried through with references to older PRs and summaries of work leading up to this. [The docs](https://docs.cilium.io/en/v1.13/contributing/development/contributing_guide/#handling-large-pull-requests) consider >200LOC and/or >6 commits to be large - this is substantially larger.

zacharysarah commented on Documentation/operations/upgrade.rst:
```diff
@@ -403,6 +403,8 @@ Helm Options
   values are kept, ``ipam.operator.clusterPoolIPv4PodCIDRList`` now defaults to
   a singleton containing the default CIDR value for the removed value
   ``ipam.operator.clusterPoolIPv4PodCIDR`` (and similarly for IPv6).
+* Values ``proxy.prometheus.enabled`` & ``proxy.prometheus.port`` are deprecated in favor of
```
nit
```suggestion
* Values ``proxy.prometheus.enabled`` and ``proxy.prometheus.port`` are deprecated in favor of
```

mhofstetter replied:
@zacharysarah thanks - fixed in the latest update.

and i agree on providing more docs as soon as we have more information.

zacharysarah reviewed (commented):
@mhofstetter What's here LGTM with one tiny nit. 👍🏻 I'm concerned that the docs impact of these changes isn't addressed here. Ideally I'd like to see updated docs included in this PR, but I'm willing to approve for docs_structure If you add them as a _very_ fast follow.

Does that sound reasonable?

bimmlerd reviewed (approved):
looked primarily at contributing stuff, lgtm

squeed commented on Documentation/observability/metrics.rst:
```diff
@@ -48,7 +48,7 @@ use ``operator.prometheus.enabled=true``.
 
 The ports can be configured via ``prometheus.port``,
-``proxy.prometheus.port``, or ``operator.prometheus.port`` respectively.
+``envoy.prometheus.port``, or ``operator.prometheus.port`` respectively.
```
Why change this? It seems reasonable enough to keep as-is...

mhofstetter replied:
@squeed the change from `proxy` to `envoy` here at this specific place in the docs or in general?

Overall it's kind of the chance to name Envoy explicitly and cleanup the naming mess with other (L7-)proxies - and therefore moving the small existing `proxy.prometheus.*` part too

squeed replied:
gotcha, makes sense.

squeed commented on install/kubernetes/cilium/templates/cilium-envoy/daemonset.yaml:
```diff
@@ -0,0 +1,254 @@
+        {{- end }}
+        livenessProbe:
+          httpGet:
+            host: {{ .Values.ipv4.enabled | ternary "127.0.0.1" "::1" | quote }}
```
Can you just use `localhost` here?

mhofstetter replied:
@squeed should be possible - changed

squeed commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -1699,13 +1704,204 @@ dashboards:
+  # Location of unix sockets for communication between Agent & Envoy.
+  socketDir:
+    # Socket directory within the envoy & agent containers
+    containers: "/var/run/cilium/envoy/sockets"
```
Why should this be configurable? Would this ever be changed?

mhofstetter replied:
@squeed within the agent - we're building on top of the `runDir` when creating the sockets dir - which unfortunately is configurable. 

e.g. if someone is changing this he should have an option to change it here as well to reflect it in the external DaemonSet.

jrajahalme replied:
This path should be a fixed path on the containter side. And on the host side this should be relative to the agent's `runDir`. This way no new helm values are needed for these paths.

mhofstetter replied:
@jrajahalme currently we only have the `runDir` - which is configured via the agent config property `--state-dir`. This only reflects the directory from within the container. a corresponding helm property is missing.

we could introduce a helm property for this and get rid of `envoy.socketDir.containers`.

but we would still lack a property for the configuration of the socket directory on the host - which is the important one IMO. reusing the same property for container & host perspective would be kind of misleading :thinking:

mhofstetter replied:
@squeed @jrajahalme found the helm property -> `daemon.runPath` -> it's only used for mounting the runDir to the host. and it's not used for the runDir within the agent via the configmap. quite confusing.

anyway. i removed the newly introduced socketDir.* variables and instead hardcoded the path within the container (which doesn't need to change) and using the `daemon.runPath` (+ relative path `/envoy/sockets`) as host mount path.

squeed commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -1699,13 +1704,204 @@ dashboards:
+    rollingUpdate:
+      maxUnavailable: 2
+  # -- Roll out cilium envoy pods automatically when configmap is updated.
+  rollOutPods: false
```
It doesn't seem like this even consumes the configmap, so this can be dropped.

squeed replied:
nevermind!

squeed commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -1699,13 +1704,204 @@ dashboards:
+
+  # -- Affinity for cilium-envoy.
+  affinity:
+    podAntiAffinity:
```
This is a daemonset; there's no need for anti-affinity.

mhofstetter replied:
@squeed `affinity` is supported by the `DaemonSet` too - and again already supported for the Cilium Agent DaemonSet.

squeed replied:
right, but as a daemonset, it already has one-per-node anti-affinity implicitly baked in. There's no need to specify any more here.

squeed replied:
We really don't want people specifying affinity here if possible, since we might want to use it to exclude nodes with embedded envoy enabled :-). I realize that rollout is later, but let's leave the avenue open.

squeed replied:
The default should still be empty; the antiAffinity here is just one-per-node, which the DaemonSet controller already does by default.

Plus, later, we can use the affinity here to control rollout from embedded to non-embedded. Just need a few cute label tricks :-)

aanm replied:
@squeed the antiAffinity is there because the DaemonSet doesn't guarantee that there will be only one pod per node. It was added by this PR https://github.com/cilium/cilium/pull/11830. IIRC I think this happened during an upgrade under specific certain conditions.

squeed replied:
wow, that's some subtle madness.

squeed commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -1699,13 +1704,204 @@ dashboards:
+    useDigest: true
+
+  # -- Additional containers added to the cilium Envoy DaemonSet.
+  extraContainers: []
```
Can we get rid of as many "extra$FOO" options as possible? Once shipped, they're impossible to remove.

mhofstetter replied:
@squeed good point. but they are kind of already shipped: it's already possible to configure them for the `cilium` (agent) DaemonSet - where the Envoy L7 Proxy was part of.

squeed replied:
Mmm, I see what you're saying, but unless we know of a good use case for extraContainers et al. with Envoy, then I'd vote for "remove first, wait for someone to complain". If it somehow breaks their workflow they can stick with embedded until we add what's needed.

mhofstetter replied:
@squeed "unfortunately" there are use cases which might be reasons to keep them.

* args & env: there are several command line options provided by Envoy. they might be used by users and or us (e.g. for analysis purposes (e.g. mentioned by @sayboras )
* volumes, mounts, ...: with the flexibility of exposing the Envoy config directly via `CiliumEnvoyConfig` users can leverage secret features (SDS) where the secrets are provided by the filesystem (also possible via environment variables) (e.g. mentioned by @jrajahalme )
* containers: might be the less used feature - but at some point its also worth to keep the configuration possibilities for the various cilium components (agent, operator, envoy, ...) aligned.

squeed commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -1699,13 +1704,204 @@ dashboards:
+    #   effect: "NoSchedule|PreferNoSchedule|NoExecute(1.6 only)"
+
+  # -- The priority class to use for cilium-envoy.
+  priorityClassName: ""
```
suggestion: set this to `~` to signify there is a default.

mhofstetter replied:
@squeed good point - changed

squeed commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -1699,13 +1704,204 @@ dashboards:
+
+  # -- DNS policy for Cilium envoy pods.
+  # Ref: https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pod-s-dns-policy
+  dnsPolicy: ""
```
likewise, should be `~`

mhofstetter replied:
@squeed there's no default for `dnsPolicy` in the envoy daemonset

squeed replied:
right, which is why this should be `~`. Otherwise `""` is documented as the default.

mhofstetter replied:
@squeed ok - changed here as well :+1:

jrajahalme commented on .github/workflows/conformance-kind-proxy-daemonset.yaml:
```diff
@@ -0,0 +1,141 @@
+
+      - name: Install Cilium
+        run: |
+          CILIUM_CLI_MODE=helm cilium install ${{ steps.vars.outputs.cilium_install_defaults }}
```
Should we install Cilium with the `external-envoy-proxy` option for that to be tested?

mhofstetter replied:
@jrajahalme cilium is already installed with the helm option `envoy.enabled=true` - which gets translated into cilium agent `external-envoy-proxy` in the config map (https://github.com/cilium/cilium/pull/25081/files#diff-83e9b8b2ccacddb2fbdb28fbcfa148473d05056c0b4c955fc971709b469d7331) 

this is exactly the intention of this GitHub Action - to test it with the external envoy (installed as DaemonSet).

jrajahalme replied:
Did not realise earlier that it was `envoy.enabled` that enables the `external-envoy-proxy` option, guess mixed it with `proxy.enabled` in my mind...

jrajahalme commented on Documentation/operations/upgrade.rst:
```diff
@@ -410,6 +410,8 @@ Helm Options
   are deprecated in favor of ``tls.ca.cert`` and ``tls.ca.key`` respectively, and
   will be removed in v1.15.
+* Values ``proxy.prometheus.enabled`` and ``proxy.prometheus.port`` are deprecated in favor of
+  their ``envoy.prometheus.*`` counterparts.
```
Typically the term "deprecation" means that the old behavior is still supported until finally removed. If we want to make a breaking change without deprecation then we should document it as such.

mhofstetter replied:
@jrajahalme the old values should still be supported (enabled & port)

jrajahalme replied:
Yes, noticed this later in the changes, forgot to edit this comment afterward, sorry.

jrajahalme commented on install/kubernetes/cilium/files/cilium-envoy/configmap/bootstrap-config.json:
```diff
@@ -0,0 +1,307 @@
+  },
+  "staticResources": {
+    "listeners": [
+      {{- if and .Values.proxy.prometheus.enabled .Values.envoy.prometheus.enabled }}
```
Should this be `or` instead of `and` ??

mhofstetter replied:
@jrajahalme `and` should be correct IMO. 

having both values a default of `true` - the `and` ensures that the deprecated value `proxy.prometheus.enabled` is still respected - e.g. a user currently sets `proxy.prometheus.enabled=false` - `or` would break this, because the default of `envoy.prometheus.enabled` is also `true`

jrajahalme replied:
Did not consider the default values, but this seems logically confusing. Luckily we get rid of this confusion when `proxy.prometheus` is removed in the release after 1.14.

jrajahalme commented on install/kubernetes/cilium/files/cilium-envoy/configmap/bootstrap-config.json:
```diff
@@ -0,0 +1,307 @@
+        "name": "envoy-health-listener",
+        "address": {
+          "socket_address": {
+            "address": "0.0.0.0",
```
Does this port need to be open on all interfaces? What about IPv6 only installations?

mhofstetter replied:
@jrajahalme thanks. it should be possible to expose health listener on loopback interfaces only.

mhofstetter replied:
@jrajahalme i changed the configuration to expose the health listener only on loopback interfaces by using the `address` & `additional_addresses` depending on whether ipv4 and/or ipv6 is enabled.

jrajahalme commented on install/kubernetes/cilium/files/cilium-envoy/configmap/bootstrap-config.json:
```diff
@@ -0,0 +1,307 @@
+        "name": "envoy-prometheus-metrics-listener",
+        "address": {
+          "socket_address": {
+            "address": "0.0.0.0",
```
Does this port need to be open on all interfaces, or can it be limited to localhost? What about IPv6-only installations?

mhofstetter replied:
@jrajahalme prometheus scrapes the prometheus data via the pods IP - after resolving all pod ips via the service (which needs to be headless for this purpose). therefore, i assume that this needs to be exposed on all interfaces.

otherwise - this will be addressed with https://github.com/cilium/cilium/issues/23353 anyway.

jrajahalme commented on install/kubernetes/cilium/templates/cilium-agent/service.yaml:
```diff
@@ -19,19 +19,21 @@ spec:
     targetPort: envoy-metrics
-{{- else if .Values.proxy.prometheus.enabled }}
+  {{- end }}
+{{- else if and .Values.proxy.prometheus.enabled .Values.envoy.prometheus.enabled (not .Values.envoy.enabled) }}
```
should this be `or` & `and` instead?

mhofstetter replied:
@jrajahalme  see https://github.com/cilium/cilium/pull/25081#discussion_r1187598781

jrajahalme reviewed (changes requested):
Small nits mostly, but we should get rid of the socket path helm values, both for host as container, as the paths are built as relative to `runDir` on the agent. We should treat them the same on helm, and can have the path as fixed on the container side, where we control the mount location explicitly anyway.

squeed commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -1699,13 +1704,204 @@ dashboards:
+          replacement: ${1}
+      # -- Metrics relabeling configs for the ServiceMonitor cilium-envoy
+      metricRelabelings: ~
+    port: "9964"
```
Hmm, shouldn't this be omitted and the values from `envoy.prometheus.port` used?

mhofstetter replied:
@squeed this is the value `envoy.prometheus.port`. for the sake of deprecation `proxy.prometheus.port` (&enabled) is kept with the deprecation info in the `values.yaml` and is also referenced in the code to use the deprecated values.

mhofstetter committed:
helm: use daemon.runPath for envoy socket paths

Helm property `daemon.runPath` is used for Envoy socket path mounting

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

jrajahalme reviewed (approved):
Thanks for addressing all the comments :-)

mhofstetter committed:
daemon: create envoy socket directory if missing at startup

The Envoy Socket directory is expected to be mounted as volume in cases
where Envoy is deployed as external DaemonSet.

To ease the use of cases where this isn't the case (e.g. embedded Envoy,
ClusterMesh external workload, ...) the directory gets automatically
creat

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter commented on daemon/cmd/daemon_main.go:
```diff
@@ -1238,6 +1224,12 @@ func initEnv() {
 		scopedLog.WithError(err).Fatal("Unable to restore agent asset permissions")
 	}
 
+	// Creating Envoy sockets directory for cases which doesn't provide a volume mount
```
@jrajahalme @sayboras [ci-external-workload](https://github.com/cilium/cilium/actions/runs/4934378356/jobs/8820182899) failed with the following error when trying to install cilium on the External VM in the clustermesh scenario

`Envoy: Failed to open xDS listen socket at /var/run/cilium/envoy/sockets/xds.sock" error="listen unix /var/run/cilium/envoy/sockets/xds.sock: bind: no such file or directory` 

This is due to the missing directory `/var/rtun/cilium/envoy/socket` in case no volume mount is provided.

i refactored this part so the directory gets created if it doesn't exist yet. 

in addition i changed the agent daemonset to only use a volumemount if envoy is deployed externally as daemonset. in the embedded case no volumemount (previously emptyDir) is used - and the directory automatically gets created too.

jrajahalme reviewed (approved):
Last two commits look OK to me :-)

mhofstetter committed:
helm: span preflight.enabled condition over whole cilium configmap

While introducing the mutual auth config section within the Cilium
ConfigMap, the `preflight.enabled` conditional block was no longer
covering the whole config map.

Appending the envoy config values at the end made this mistake visible -
it results in templating issues.

This commit spans the conditional block over the whole configmap.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

nathanjsweet reviewed (approved):
Overall it looks, but I have one overriding concern that needs issue documentation:

Cross-pod communication via a UNIX domain socket will trip SELinux up and not work on platforms like Openshift. The preferred method for doing this is a service (which you can secure via mTLS if needed) with `internalTrafficPolicy: Local` set, which ensures that Envoy and Cilium will always communicate on the local node. See [this PR](https://github.com/cilium/cilium/pull/18620) as an example conversion.

It's fine that this PR doesn't do this, but we must document this limitation.

A follow on Documentation PR along with an issue to fix should be good.
